satellite/gracefulexit/endpoint.go

```diff
@@ -0,0 +1,390 @@
+					endpoint.log.Debug("no more pieces to transfer for node.", zap.String("node ID", nodeID.String()))
+					atomic.StoreInt32(&morePiecesFlag, 0)
+					break
+				}
+
+				for _, inc := range incomplete {
+					pointer, err := endpoint.metainfo.Get(ctx, string(inc.Path))
+					if err != nil {
+						return Error.Wrap(err)
+					}
+					remote := pointer.GetRemote()
+
+					found := false
+					for _, piece := range remote.GetRemotePieces() {
+						if piece.NodeId == nodeID && piece.PieceNum == inc.PieceNum {
+							found = true
+						}
+					}
+					if !found {
+						endpoint.log.Debug("piece no longer held by node.", zap.String("node ID", nodeID.String()), zap.ByteString("path", inc.Path), zap.Int32("piece num", inc.PieceNum))
+						continue
+					}
+
+					redundancy, err := eestream.NewRedundancyStrategyFromProto(pointer.GetRemote().GetRedundancy())
+					if err != nil {
+						return Error.Wrap(err)
+					}
+					pieceSize := eestream.CalcPieceSize(pointer.GetSegmentSize(), redundancy)
+
+					request := overlay.FindStorageNodesRequest{
+						RequestedCount: 1,
+						FreeBandwidth:  pieceSize,
+						FreeDisk:       pieceSize,
+						ExcludedNodes:  []storj.NodeID{nodeID},
+					}
+
+					newNodes, err := endpoint.overlay.FindStorageNodes(ctx, request)
+					if err != nil {
+						return Error.Wrap(err)
+					}
+
+					if len(newNodes) == 0 {
+						return Error.New("could not find a node to transfer this piece to. nodeID %v, path %v, pieceNum %v.", nodeID.String(), zap.ByteString("path", inc.Path), inc.PieceNum)
+					}
+
+					endpoint.log.Debug("found new node for piece transfer.", zap.String("node ID", newNodes[0].Id.String()), zap.ByteString("path", inc.Path), zap.Int32("piece num", inc.PieceNum))
+
+					parts := storj.SplitPath(string(inc.Path))
+					if len(parts) < 2 {
+						return Error.New("invalid path %v.", zap.ByteString("path", inc.Path))
+					}
+
+					bucketID := []byte(storj.JoinPaths(parts[0], parts[1]))
+					limit, privateKey, err := endpoint.orders.CreateGracefulExitPutOrderLimit(ctx, bucketID, newNodes[0].Id, inc.PieceNum, remote.RootPieceId, remote.Redundancy.GetErasureShareSize())
+					if err != nil {
+						return Error.Wrap(err)
+					}
+
+					transferMsg := &pb.SatelliteMessage{
+						Message: &pb.SatelliteMessage_TransferPiece{
+							TransferPiece: &pb.TransferPiece{
+								PieceId:             limit.Limit.PieceId,
+								AddressedOrderLimit: limit,
+								PrivateKey:          privateKey,
+							},
+						},
+					}
+					err = stream.Send(transferMsg)
+					if err != nil {
+						return Error.Wrap(err)
+					}
+					pending.Store(limit.Limit.PieceId, pendingTransfer{
+						path:             inc.Path,
+						pieceSize:        pieceSize,
+						satelliteMessage: transferMsg,
+					})
+				}
+			}
+		}
+		return nil
+	})
+
+	for {
+		// if there are no more transfers and the pending queue is empty, send complete
+		if atomic.LoadInt32(&morePiecesFlag) == 0 && pendingLength() == 0 {
```

Comment:
Save the value of `pendingLength()` as a variable so that we do not need to call it again below

Reply:
Updated